fv3core/decorators.py

```diff
@@ -150,16 +150,169 @@ def __iter__(self):
+        domain: Optional[Index3D] = None,
+        validate_args: Optional[bool] = None,
+        **kwargs,
+    ) -> None:
+        if self.origin:
+            assert origin is None, "cannot override origin provided at init"
+            origin = self.origin
+        if not self.field_origins or self.disable_cache:
+            self.field_origins = self._compute_field_origins(origin, *args, **kwargs)
+
+        if self.domain:
+            assert domain is None, "cannot override domain provided at init"
+            domain = self.domain
+        else:
+            assert domain is not None, "no domain provided at call time"
+
+        if not validate_args:
+            validate_args = global_config.get_validate_args()
+
+        if validate_args:
+            self.stencil_object(
+                *args,
+                **kwargs,
+                origin=self.field_origins,
+                domain=domain,
+                validate_args=True,
+            )
+        else:
+            kwargs = self._process_kwargs(domain, *args, **kwargs)
+            self.stencil_object.run(**kwargs, exec_info=None)
+
+    def _process_kwargs(self, domain: Optional[Index3D], *args, **kwargs):
+        """Processes keyword args for direct calls to stencil_object.run."""
+
+        kwargs["_origin_"] = self.field_origins
+        kwargs["_domain_"] = domain
+
+        arg_names = self.field_names + self.parameter_names
+        for i in range(len(args)):
+            kwargs[arg_names[i]] = args[i]
+
+        return kwargs
+
+    def _compute_field_origins(
+        self, origin: Tuple[int, ...], *args, **kwargs
+    ) -> Dict[str, Tuple[int, ...]]:
+        """Computes the origin for each field in the stencil call."""
+
+        field_origins: Dict[str, Tuple[int, ...]] = {"_all_": origin}
+        field_names: List[str] = self.field_names
+        for i in range(len(field_names)):
+            field_name = field_names[i]
+            field_axes = (
+                self.stencil_object.field_info[field_name].axes
+                if self.stencil_object.field_info[field_name]
+                else []
+            )
+            field_shape = args[i].shape if i < len(args) else kwargs[field_name].shape
+            if field_axes == ["K"]:
+                field_origin = [min(field_shape[0] - 1, origin[2])]
+            else:
+                field_origin = [
+                    min(field_shape[j] - 1, origin[j]) for j in range(len(field_shape))
+                ]
+            field_origins[field_name] = tuple(field_origin)
+        return field_origins
+
+    def _set_device_sync(self, stencil_kwargs: Dict[str, Any]):
+        """Sets the 'device_sync' backend option in the stencil kwargs."""
+        if self.device_sync is None and "cuda" in self.backend:
+            self.device_sync = global_config.get_device_sync()
+        if self.device_sync is not None:
+            stencil_kwargs["device_sync"] = self.device_sync
+
+    @property
```

Comment:
if we make these two properties, should we also create proper setters?

Reply:
These are read-only convenience functions to extract information from the `stencil_object`, so I would say no setters.